ratelimits/limiter.go

```diff
@@ -0,0 +1,230 @@
+package ratelimits
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/jmhodges/clock"
+)
+
+// ErrInvalidCost indicates that the cost specified was <= 0.
+var ErrInvalidCost = fmt.Errorf("invalid cost, must be > 0")
+
+// ErrInvalidCostForCheck indicates that the check cost specified was < 0.
+var ErrInvalidCostForCheck = fmt.Errorf("invalid check cost, must be >= 0")
+
+// ErrInvalidCostOverLimit indicates that the cost specified was > limit.Burst.
+var ErrInvalidCostOverLimit = fmt.Errorf("invalid cost, must be <= limit.Burst")
+
+var ErrBucketAlreadyFull = fmt.Errorf("bucket already full")
+
+// Limiter provides a high-level interface for rate limiting requests by
+// utilizing a leaky bucket-style approach.
+type Limiter struct {
+	// defaults stores default limits by 'name'.
+	defaults limits
+
+	// overrides stores override limits by 'name:id'.
+	overrides limits
+
+	// source is used to store buckets. It must be safe for concurrent use.
+	source source
+	clk    clock.Clock
+}
+
+// NewLimiter returns a new *Limiter. The provided source must be safe for
+// concurrent use. The defaults and overrides paths are expected to be paths to
+// YAML files that contain the default and override limits, respectively. The
+// overrides file is optional, all other arguments are required.
+func NewLimiter(clk clock.Clock, source source, defaults, overrides string) (*Limiter, error) {
+	limiter := &Limiter{source: source, clk: clk}
+
+	var err error
+	limiter.defaults, err = loadLimits(defaults)
+	if err != nil {
+		return nil, err
+	}
+
+	if overrides == "" {
+		// No overrides specified.
+		limiter.overrides = make(limits)
+		return limiter, nil
+	}
+
+	limiter.overrides, err = loadLimits(overrides)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
```suggestion
	if overrides != "" {
		limiter.overrides, err = loadLimits(overrides)
		if err != nil {
			return nil, err
		}
	}
```

Reply:
We want to initialize this mapping regardless to avoid attempting to check a nil map when we call getLimit()

Reply:
I'll leave a comment to indicate this.